Celery/ Celery-beat

CeNTePeDe opened the pull request.

CeNTePeDe committed:
added items to jsonfield_car_showroom
removed method clean from CarShowRoom
added to CarProvider model:
- discount_provider
removed from Transaction:
- discount
- season_discount
added to BaseSeLLModel:
- season_discount
- discount
create task for car_showroom
added discount for car
override:
- get_season_discount
- get_provider_discount
created discount function
created task find cars
created property for provider,
added choices for CarShowRoom validators
remove function from constants
added date for sell_model
created func for car_showroom characteristic:
- get_color
- get_number_of_door
- get_engine_type
- get_body_type
- jsonfield_provider
added clean method to CarShowRoom Model
rename validators_show_room -> validators
created jsonfield for customer
creates validators for jsonfield
created func:
- get_car_list()
- get_car_models_list()
- jsonfield_customer()
added characteristic_car for admin-customer pannel
override  method clean for customer model(check max_price)
added method calculate_price  for provider_price
add celery and celery-beat
removed characteristic from provider model

slavadan commented on show_room_app/core/constants.py:
```diff
@@ -0,0 +1,38 @@
+
+
+def get_engine_type():
+    return [choice.value for choice in EnginType] + [""]
```
Engine

slavadan commented on show_room_app/customer/models.py:
```diff
@@ -23,6 +30,16 @@ class Customer(models.Model):
+            self.characteristic_car["name"]
+        ):
+            raise ValidationError(_("Car model does not exist"))
+
```
Если честно я не понимаю смысл этих валидаров тк если машин с такими характеристиками не существуют они просто не найдуться и все. А если цена машины больше чем баланс то автосалон просто не сможет ее купить. Но хвалю за то что нашла метод для валидации филдов. Короче удали валидаторы и get_car_list и другие методы в constants, которые могут возвращать разные результаты.

slavadan commented on show_room_app/core/constants.py:
```diff
@@ -0,0 +1,38 @@
+def get_car_models_list(car_name):
+    cars = Car.objects.filter(name=car_name)
+    models = [car.model_car for car in cars]
+    return models + [""]
```
Я бы не назвал get_car_list и get_car_models_list константными тк модели добавляются постоянно они будут постоянно изменяться. Насчет чойсов тоже самое хотя они изменяются довольно рндко .  Вот jsonfield_customer это да они всегда возвращают одно и тоже.

slavadan commented on show_room_app/provider/models.py:
```diff
@@ -22,8 +23,28 @@ class CarProvider(BaseSellModel):
+
+    def save(self, *args, **kwargs):
+        self.provider_price = self.calculate_price()
+        super(CarProvider, self).save(*args, **kwargs)
```
Довольно спорный момент. Тк с одной стороны provider_price позволит отображать цену в бд и джанго админке. С другой стороны при малейшем изменении модели даже если менялась не цена, то будет постоянно пересчитыватьтся, а это доп нагрузка. Я бы удалил поле provider_price, calculate_price сделал бы @property price, удалил бы save(). А если хочешь отобразить цену в джанго админке то в классе-наследнике от modeladmin сделал бы @attributed метод и прокинул бы его в list_display, чтобы она отображалась на странице со списком всех машин которые продают все поставищики, а чтобы цена отображалась на форме в админке, то переопредил поле fields.

slavadan commented on show_room_app/car_showroom/models.py:
```diff
@@ -14,6 +25,16 @@ class CarShowRoom(BaseRole):
+        if self.characteristic["number_of_doors"] not in get_number_of_door():
+            raise ValidationError(_("Choose the correct number of door"))
+        if self.characteristic["body_type"] not in get_body_type():
+            raise ValidationError(_("this body type does not exist"))
```
Если честно я не понимаю смысл этих валидаров тк если машин с такими характеристиками не существуют они просто не найдуться и все. А если цена машины больше чем баланс то автосалон просто не сможет ее купить. Но хвалю за то что нашла метод для валидации филдов. Короче удали валидаторы и get_car_list и другие методы в constants, которые могут возвращать разные результаты.

slavadan commented on show_room_app/car_showroom/tasks.py:
```diff
@@ -0,0 +1,32 @@
+    for provider in car_providers:
+
+        sell_model = SellModel.objects.create(
+            car_showroom=CarShowRoom.objects.get(user_id=user_id),
```
ну камон два раза гет прокидывать, ну ты чо. Сделай один гет и доставай все что тебе нужно. Убери пустую строку между for и sell_model

slavadan commented on show_room_app/car_showroom/tasks.py:
```diff
@@ -0,0 +1,32 @@
+            car=provider.car,
+            provider=provider.provider,
+            discount=get_provider_discount(provider_id=provider.provider.user_id, car_showroom_id=user_id),
+            season_discount=get_season_discount(sale_day=date.today()),
```
а вот как ты решила делать

slavadan replied:
get_provider у тебя скидка, которая работает при покупки некоторого кодичества машин, а чо ты будешь делать если на машину тоже есть обычная скидка????

slavadan replied:
кстати а что если автосалон захочет купить 10 машин за раз?

slavadan replied:
+ где проверка на то, что если у автсалона/кастомера нету денег на балансе

slavadan replied:
кстати еще интересный момент, а где снятие денег со счета

slavadan commented on show_room_app/car_showroom/tasks.py:
```diff
@@ -0,0 +1,32 @@
+            season_discount=get_season_discount(sale_day=date.today()),
+
+        )
+        sell_model.save()
```
Да емае ну 3й раз на одни и теже грабли. Ну сколько можно

slavadan commented on show_room_app/core/services.py:
```diff
@@ -0,0 +1,33 @@
+    for car_price in cars.provider_price:
+        final_provider_price = (get_provider_discount(provider_id, car_showroom_id).discount_rate + get_season_discount(
+            sale_day=date.today()).discount_rate)/100*car_price + car_price
+        return final_provider_price
```
это вообще как будет работать? зачем поле на всю сумму денег, ну мыже договорились его убрать

slavadan replied:
у тебя же есть проперти

slavadan replied:
удаляй

slavadan replied:
Тайп хинтов нету, а я говорил их добавить

CeNTePeDe committed:
added to BaseDiscount fields date_start, date_finish
removed SeasonDiscount
created list of json for car_characteristic

CeNTePeDe committed:
created task buy_car_from_provider
added func for discount
Added docstring to models
Removed transaction field in Customer model
Added customer field(FK to customer) to Transaction model

CeNTePeDe committed:
calculate balance for car_showroom
override serializer
created task for customer
added admin panel for discount
created file car_showroom_discount for calculate discount car_showroom
removed calculate for discount
added func get_full_discount
created method get_full_discount
added docstring for methods:
- get_season_discount_from_provider
- get_provider_discount_for_special_client
- get_full_discount
renamed field characteristic to list_of_car_model

slavadan commented on show_room_app/core/car_showroom_discount.py:
```diff
@@ -0,0 +1,76 @@
+        + get_car_showroom_discount_for_special_client(
+            car_showroom_id=car_showroom_id, customer_id=customer_id
+        )
+    ) / 100
```
все еще не читабельно, говорил же, разбей по строчкам

slavadan commented on show_room_app/core/car_showroom_discount.py:
```diff
@@ -0,0 +1,76 @@
+                discount_name="first_purchase_discount", is_active=True
+            ).discount_rate
+        elif 0 > count_car_customer < 2:
+            return 0
```
лишний elif

slavadan commented on show_room_app/car_showroom/tasks.py:
```diff
@@ -0,0 +1,59 @@
+
+
+@shared_task()
+def buy_car_from_provider(user_id):
```
блин, ну я уже 3й день говорю про тайп хинты. Вчера даже сказал пометь себе. Меня это уже бесить начинает довольно сильно

slavadan commented on show_room_app/car_showroom/tasks.py:
```diff
@@ -0,0 +1,59 @@
+        None
+    """
+    car_showroom = CarShowRoom.objects.get(user_id=user_id)
+    list_of_cars = car_showroom.list_of_car_model["cars"]
```
`list_of_car_model` никак не отображает суть для чего нужна эта переменная. Мы же вчера сидели и я говорил примерное название, которое отображает суть этого json, также сказал что можно другое название, но главное чтобы суть была ясна

slavadan commented on show_room_app/car_showroom/tasks.py:
```diff
@@ -0,0 +1,59 @@
+            # add similar car to one dictionary car_with_price
+            car_with_price.update({car_item: price})
+        # get min price for car
+        car_min_price = min(car_with_price.items(), key=lambda x: x[1])
```
все еще максимально не читабельно. Вчера уже говорил как можно было бы сделать

slavadan commented on show_room_app/core/car_showroom_discount.py:
```diff
@@ -0,0 +1,76 @@
+        car_showroom_id(int): Id car_showroom.
+
+    Returns:
+        discount_rate(int): Discount_rate for particular customer
```
Неправильно документируешь функцию, другие тоже

slavadan commented on show_room_app/core/provider_discount.py:
```diff
@@ -0,0 +1,83 @@
+        )
+        return provider.discounts.get(
+            discount_name="season_discount", is_active=True
+        ).discount_rate
```
ясно, вчера созвон частично прошел впустую. Все же расписали

slavadan commented on show_room_app/core/provider_discount.py:
```diff
@@ -0,0 +1,83 @@
+    else:
+        return provider.discounts.get(
+            discount_name="no_discount", is_active=True
+        ).discount_rate
```
тоже проговаривали

slavadan commented on show_room_app/core/provider_discount.py:
```diff
@@ -0,0 +1,83 @@
+        )
+    ) / 100
+
+    return discount
```
все еще не читабельно, лучше не стало. Вчера проговаривали

slavadan commented on show_room_app/customer/tasks.py:
```diff
@@ -0,0 +1,57 @@
+            )
+            + car_item.price
+        )
+        car_with_price.update({car_item: price})
```
Не читабельно

slavadan commented on show_room_app/customer/tasks.py:
```diff
@@ -0,0 +1,57 @@
+        None
+    """
+    customer = Customer.objects.get(user_id=user_id)
+    model_car = customer.characteristic_car
```
а в этой модели переименовать поле?

slavadan commented on show_room_app/customer/tasks.py:
```diff
@@ -0,0 +1,57 @@
+            customer.save()
+
+            # add price_car to car_showroom balance
+            car_showroom.balance += car_min_price[1].amount
```
А сохранить?

CeNTePeDe committed:
added try/except block for each discount
rename characteristic to model_car for customer
added filter to admin for discount
override tasks for car_showroom and provider
added filter to admin fo discount

CeNTePeDe committed:
created schedule for buy_car_from_provider
created signal for customer app for task
added func get_user_id_for_car_showroom
created api_client:
- simple_api_client
- admin_api_client
- customer_api_client
- car_showroom_api_client
created filter:
- SellModel
- CarProvider
add celery-beat

CeNTePeDe committed:
added html_file:
- reset_password_email.html
- activation_email.html
added for users:
- RegistrationAPIView
- ForgotPasswordView
- ResetPasswordView
- UserLoginAPIView
- UserLogoutAPIView
added settings for sending email
added filter to SellModel, Transaction
created tasks buy_car_from_provider_for_each_car_showroom

slavadan commented on show_room_app/car_showroom/tasks.py:
```diff
@@ -6,54 +6,55 @@
+def buy_car_from_provider_for_each_car_showroom() -> None:
+    user_ids = list(CarShowRoom.objects.values_list("user_id", flat=True))
+    for user_id in user_ids:
+        buy_car_from_provider.delay(user_id)
```
values_list и list()

slavadan commented on show_room_app/user/serializer.py:
```diff
@@ -73,15 +74,22 @@ def create(self, validated_data):
+                setattr(instance, key, value)
+        instance.save()
+        return instance
+
```
`
if validated_data.get('password'):
    instance.set_password(validated_data.pop('password')
super().update(instance, validated_data)
`

переопределяй правильно

slavadan commented on show_room_app/user/serializer.py:
```diff
@@ -91,36 +99,62 @@ def validate(self, data):
     def update(self, instance, validated_data):
         instance.set_password(validated_data["password"])
+        instance.set_password(validated_data["password_confirm"])
         instance.save()
```
ты чооооо. зачем хранить в бд потдверждение пароля?

slavadan commented on show_room_app/user/views.py:
```diff
@@ -36,22 +45,98 @@ def retrieve(self, request, pk=id, *args, **kwargs):
+                "token": default_token_generator.make_token(user),
+            },
+        )
+        send_mail(subject, message, settings.EMAIL_HOST_USER, [user.email])
```
переделай не в perform_create(), и без 2 save()

slavadan commented on show_room_app/user/views.py:
```diff
@@ -36,22 +45,98 @@ def retrieve(self, request, pk=id, *args, **kwargs):
+            },
+        )
+        send_mail(subject, message, settings.EMAIL_HOST_USER, [user.email])
+        return Response({"status": "email sent"})
```
Не помню, чтобы среди 100ых, 200ых статус кодов существовал 'semail sent'. Переименуй,

slavadan commented on show_room_app/user/views.py:
```diff
@@ -36,22 +45,98 @@ def retrieve(self, request, pk=id, *args, **kwargs):
+            return Response({"status": "activated"})
+        else:
+            return Response(
+                {"status": "invalid token"}, status=status.HTTP_400_BAD_REQUEST
```
два статуса Xd

slavadan commented on show_room_app/user/views.py:
```diff
@@ -36,22 +45,98 @@ def retrieve(self, request, pk=id, *args, **kwargs):
+            return Response({"status": "password reset"})
+        else:
+            return Response(
+                {"status": "invalid token"}, status=status.HTTP_400_BAD_REQUEST
```
два статуса

CeNTePeDe committed:
task buy_car_from_provider_for_each_car_showroom:
- removed list

User model
- removed password_confirm
- override is_active field (default=False)
UserRegistrationSerializer:
- add validator for base_role
- add UniqueValidator for username field
UserManage
- removed validator for base_role
ActivateView:
- rename response to message
RegistrationAPIView:
- override create method

CeNTePeDe committed:
Discounts:
- remove season_discount
- override provider_discount
- override test_car_showroom_discount

CeNTePeDe committed:
Discounts:
- remove season_discount
- override provider_discount
- override test_car_showroom_discount
DiscountFactory added:
- date_start
- discount_rate
- special_discount
- is_active

CeNTePeDe committed:
ProviderFactory:
- remove characteristic
test_provider added:
- simple_api_client

CeNTePeDe committed:
created test for get_season_discount_from_provider:
-test_get_season_discount_from_provider
-test_get_season_discount_from_provider_no_discount

created test for get_season_discount_from_car_showroom:
-test_get_season_discount_from_car_showroom
-test_get_season_discount_from_car_showroom_no_discount

TransactionSerializer:
- added SlugRelatedField
CarProviderSerializer:
- added SlugRelatedField foe provider
added pytest-celery
added validator for moneyfield

Discount:
- add permissions to view
- override tests
Permission:
- add permissions IsCarShowroomDiscountOrReadOnly, IsProviderDiscountOrReadOnly
CarShowRoomDiscount:
- added simple_api_client
- force_authenticate

Test car_provider:
- added update method to serializer

Test Customer:
- added token authenticate
CustomerSerializer:
- remove methods create

Test CarShowRoom
- added force_authenticate for simple_api_client
CarShowRoomFactory:
- rename characteristic to list_cars_to_buy
CarShowRoomSerializer:
- remove method create

Test CarProvider:
- added dict for car
CarProviderSerializer:
- added create method
build_car_provider:
- removed car, provider
CarProviderFactory:
- removed number_of_cars

Provider serializer:
- remove method create
conftest.py
- removed api_clients for roles

Created tests for user:
- test_create_user_customer
- test_confirm_user_customer
- test_create_user_car_showroom
- test_confirm_user_car_showroom
- test_create_user_provider
- test_confirm_user_provider

rename fields

CeNTePeDe committed:
created fixtures for celery
created tasks for celery-worker

created function for buy_car_from_car_showroom tasks:
- matching_car_for_customer
- find_car_with_min_price
- purchase_marching_car_for_customer
created test Provider module get_provider_discount_for_special_client:
- test_provider_discount_for_first_purchase
- test_provider_discount_no_discount
- test_provider_discount_regular_discount

created test for get_car_showroom_discount_for_special_client:
- test_car_showroom_discount_for_first_purchase
- test_car_showroom_discount_no_discount
- test_car_showroom_discount_regular_customer

CeNTePeDe committed:
added default.conf file for nginx
added nginx container
added gunicorn

slavadan committed:
Merge pull request #7 from CeNTePeDe/feature/tests

Tests for show_room_app

slavadan committed:
Merge pull request #8 from CeNTePeDe/feature/gunicorn_nginx

add nginx and gunicorn
